spi/imx6ull-qspi/qspi.c

```diff
@@ -0,0 +1,326 @@
+/*
+ * Phoenix-RTOS
+ *
+ * i.MX 6ULL QSPI lib
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Hubert Badocha
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <stddef.h>
+#include <sys/interrupt.h>
+#include <sys/mman.h>
+#include <errno.h>
+#include <sys/platform.h>
+#include <sys/threads.h>
+#include <stdbool.h>
+#include <string.h>
+#include <phoenix/arch/imx6ull.h>
+#include "qspi.h"
+#include <board_config.h>
+
+/* Default values for not configured boards */
+#ifndef QSPI_FLASH_A1_SIZE
+#define QSPI_FLASH_A1_SIZE (256 * 1024 * 1024)
+#endif
+#ifndef QSPI_FLASH_A2_SIZE
+#define QSPI_FLASH_A2_SIZE (0)
+#endif
+#ifndef QSPI_FLASH_B1_SIZE
+#define QSPI_FLASH_B1_SIZE (0)
+#endif
+#ifndef QSPI_FLASH_B2_SIZE
+#define QSPI_FLASH_B2_SIZE (0)
+#endif
+
+#define QSPI_MMAP_BASE 0x60000000
+
+#define QSPI_BASE ((uint32_t *)0x21E0000)
+
+#define QSPI_LCKCR_ADDR   (qspi_common.base + 0x304 / sizeof(uint32_t))
+#define QSPI_LCKCR_LOCK   0x01
+#define QSPI_LCKCR_UNLOCK 0x02
+
+#define QSPI_LUT_KEY_ADDR (qspi_common.base + 0x300 / sizeof(uint32_t))
+#define QSPI_LUT_KEY      0x5AF05AF0
+
+#define QSPI_LUTn_ADDR(n)      (qspi_common.base + (0x310 + 4 * (n)) / sizeof(uint32_t))
+#define QSPI_SEQ_START_REGN(n) (4 * (n))
+
+#define QSPI_IPCR_ADDR                    (qspi_common.base + 0x08 / sizeof(uint32_t))
+#define QSPI_IPCR_SET_SEQID(reg, seqid)   ((reg) & ~(0x0F << 24)) | ((seqid & 0x0F) << 24)
+#define QSPI_IPCR_SET_IDATSZ(reg, idatsz) ((reg) & ~(0xFFFF)) | (idatsz)
```

Comment:
We strive to reduce number of macros to the minimum. These are only used once, so there is no incentive to keep them
```suggestion
```